Task solved according to checklist.

fedaktaras opened the pull request:
- tried to move if/else logic away or use ternary but not too much, just to keep it readable.

fedaktaras committed:
Task solved according to checklist.
- tried to move if/else logic away or use ternary but not too much, just to keep it readable.

TonyH277 commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+        return x;
+    }
+
+    private void initializeBase(Node<T> node) {
```
Minor. Just for clarity as you defined startNode and endNode

```suggestion
    private void initializeStartNode(Node<T> node) {
```

TonyH277 commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+        size++;
+    }
+
+    private void initializeTale(Node<T> node) {
```
```suggestion
    private void initializeEndNode(Node<T> node) {
```

TonyH277 commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
         return false;
     }
+
+    private class Node<T> {
```
it can be made private static to clearly indicate its relationship with MyLinkedList

TonyH277 reviewed (changes requested):
Really good job, request changes so you could edit some mistakes -> get aproved faster🥇

fedaktaras committed:
Sent for mentors check by accident. On Q&A at the moment. I know I should refactor to Node node = new node(value, Next, Prev)

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+        }
+    }
+
+    private Node<T> findNodeFromStart(int index) {
```
```suggestion
    private Node<T> findNodeFromHead(int index) {
```

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> startNode;
```
```suggestion
    private Node<T> head;
```

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> startNode;
+    private Node<T> endNode;
```
```suggestion
    private Node<T> tail;
```

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+    }
+
+    private Node<T> findNodeFromStart(int index) {
+        Node<T> x = startNode;
```
```suggestion
        Node<T> currentNode = startNode;
```

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+    }
+
+    private Node<T> findNodeFromEnd(int index) {
+        Node<T> x = endNode;
```
```suggestion
        Node<T> currentNode = endNode;
```

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+    }
+
+    private void indexValidation(int index) {
+        if (index < size && index >= 0) {
+            return;
+        }
+        throw new IndexOutOfBoundsException("Index " + index + " out of bounds.");
```
```suggestion
        if (index > size && index <= 0) {
            throw new IndexOutOfBoundsException("Index " + index + " out of bounds.");
        }
        
```

karma-o replied:
not resolved

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+        throw new IndexOutOfBoundsException("Index " + index + " out of bounds.");
+    }
+
+    private boolean addedWithBaseOrTaleInitialization(T value, int index) {
+        if (size == 0 && index == 0) {
+            initializeStartNode(new Node<>(value));
+            return true;
+        } else if (size == 1 && index == 1) {
+            initializeEndNode(new Node<>(value));
+            return true;
+        }
+        return false;
+    }
+
+    private boolean addedWithBaseOrTaleInitialization(T value) {
+        if (size == 0) {
+            initializeStartNode(new Node<>(value));
+            return true;
+        } else if (size == 1) {
+            initializeEndNode(new Node<>(value));
+            return true;
+        }
         return false;
     }
```
These methods used only once and significantly increase complexity, so I suggest creating methods addToHead and addToTail and use them where you need

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+    }
+
+    private void linkFirst(Node<T> node) {
+        startNode.previous = node;
```
You have kind of 2 methods to set head linkFirst and initializeStart node i suggest to collaplse them (as well as linkLast and initializeEnd)

andrii-hoienko commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,236 @@
+
+    private T removeByIndex(int index) {
+        Node<T> node = findByIndex(index);
+        if (size == 2) {
+            if (index == 1) {
+                startNode.next = null;
+                endNode = null;
+            } else if (index == 0) {
+                startNode = endNode;
+                endNode = null;
+                startNode.next = null;
+            }
+        } else if (size == 1) {
+            node = startNode;
+            startNode = null;
+        } else {
+            node.previous.next = node.next;
+            node.next.previous = node.previous;
+        }
```
I believe here you should just call unlink

andrii-hoienko reviewed (changes requested):
Good job, but some improvements required

fedaktaras committed:
Refactored code.
It was difficult to refactor.
Almost Everything has been changed.
But now as much logic is encapsulated as I could do.
- Node can link itself properly depending on neighbour nodes.
- linkLast(value) and linkFirst(value) logic determines the content
- single unlink(Node)

Hope no more logic changes) It was vey difficult to refactor

karma-o commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,205 @@
+        return x;
+    }
+
+    private void indexValidation(int index) {
```
method names should be verbs
```suggestion
    private void validateIndex(int index) {
```

karma-o commented on src/main/java/core/basesyntax/MyLinkedList.java:
```diff
@@ -3,45 +3,205 @@
+        private Node<T> previous;
+        private Node<T> next;
+
+        private Node(T data, Node<T> next, Node<T> previous) {
```
constructors should not have any logic inside them, only variable assignments or calls to other constructors

fedaktaras committed:
Refactored code.
- no logic in constructor.
- validateIndex() changed
- linkLast() - removed

yevheniisynytsia reviewed (approved):
you didn't fix mentor's comments about naming, please consider this comments
it is important to learn how to name methods and variables
